### PR TITLE
Lineage support for JdbcIO

### PR DESCRIPTION
**Pull request by Abacn**
* Report table the pipeline read from and write to

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 1]
```diff
@@ -39,6 +39,7 @@
 import java.util.Collection;
```
Comment (rohitsinha54) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:39]: Should we update the documentation of the IO to specifically say the lineage metrics are only collected in these cases to set customer expectations correctly.

**Review by rohitsinha54 — APPROVED** [review]
Thank you.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcUtil.java [hunk 2]
```diff
@@ -563,4 +577,232 @@ public KV<Long, KV<DateTime, DateTime>> mapRow(ResultSet resultSet) throws Excep
...
+  @AutoValue
+  abstract static class JdbcUrl {
+    abstract String getScheme();
+
+    abstract @Nullable String getHostAndPort();
+
+    abstract String getDatabase();
+
+    static @Nullable JdbcUrl of(String url) {
+      if (Strings.isNullOrEmpty(url) || !url.startsWith("jdbc:")) {
```
[the remark this answers is not in the document]
Comment (Abacn) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcUtil.java:590]: yes from test. For HikariDataSouce, checked that the whitespace in the front does not affect the result of getJdbcUrl() call; For DataSource setup by field, it fails when host contains trailing space

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 3]
```diff
@@ -2603,11 +2624,28 @@ public void setup() {
...
+        String table = spec.getTable();
+        if (Strings.isNullOrEmpty(table) && spec.getStatement() != null) {
+          table = JdbcUtil.extractTableFromWriteQuery(spec.getStatement().get());
+        }
+        if (!Objects.equals(table, reportedLineage)) {
+          JdbcUtil.FQNComponents fqn = JdbcUtil.FQNComponents.of(validSource);
+          if (fqn == null) {
+            fqn = JdbcUtil.FQNComponents.of(connection);
+          }
+          if (fqn != null) {
```
[the remark this answers is not in the document]
Comment (Abacn) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:2644]: added

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcUtil.java [hunk 4]
```diff
@@ -563,4 +577,232 @@ public KV<Long, KV<DateTime, DateTime>> mapRow(ResultSet resultSet) throws Excep
               }
             }
           });
+
+  @AutoValue
+  abstract static class JdbcUrl {
```
[the remark this answers is not in the document]
Comment (Abacn) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcUtil.java:582]: Yeah, admit it is brittle. Added log.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java (hunk 1 above)
Comment (Abacn) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:39]: added documentations for JdbcUtil.JdbcUrl.of. Was hesitate on adding the information to top level JdbcIO javadoc, as it is usually designated for instruction for read and write transform.

**Pull request merged**